### PR TITLE
[schema] Support compatibility check over all existing schemas.

### LINKED ISSUE #4170: SchemaRegistry should support compatibility check over all existing schemas
**Is your feature request related to a problem? Please describe.**

Currently the schema compatibility check is only done with the latest schema. That's probably not enough especially we now support schema versions for Struct schema. We need to the ability for check the compatibility with all existing schema version.

**Describe the solution you'd like**

We should introduce *transitive* schema check strategy for `FORWARD`, `BACKWARD` and `FULL`.

Hence, they will be `FORWARD_TRANSITIVE`, `BACKWARD_TRANSITIVE` and `FULL_TRANSITIVE`.

**Describe alternatives you've considered**

N/A

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by codelipenghui**
Fixes #4170 

### Motivation

Currently the schema compatibility check is only done with the latest schema. That's probably not enough especially we now support schema versions for Struct schema. We need to the ability for check the compatibility with all existing schema version.

### Modifications

Add `FORWARD_TRANSITIVE`, `BACKWARD_TRANSITIVE` and `FULL_TRANSITIVE` compatibility strategy.

### Verifying this change

This change is already covered by existing tests, such as SchemaServiceTest, BaseAvroSchemaCompatibilityTest

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (yes)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by merlimat** [comment]
@codelipenghui Could this possibly lead to a situation where a schema `s1` could be updated into 2 diverging `s2` and `s3`, incompatible between each others?

**Comment by codelipenghui** [comment]
@merlimat 
Schema registry can avoid both s2 and s3 will be update successfully.
https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/BookkeeperSchemaStorage.java#L450

**Comment by merlimat** [comment]
I'm not referring to race conditions but to a much simpler case: 

 * Topic has only 1 schema `s1` 
 * Some producer connects and updates the schema by adding a field, eg: `int a`. 
 * Now we have `s2` and `s1` as schemas for the topic
 * Some other producer might come and update with a new schema, same as `s1` but with a new field `string a`. 

Note that: 
 * `s1` -> `s2` (compatible)
 * `s1` -> `s3` (compatible)
 * `s2` -> `s3` (non-compatible)

Now if we check with all the schemas how this would work? would `s3` be allowed or not?

**Comment by sijie** [comment]
s3 will be rejected if `_TRANSITIVE` checker is used.

**Comment by codelipenghui** [comment]
run java8 tests

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/DefaultSchemaRegistryService.java [hunk 1]
```diff
@@ -35,6 +36,12 @@
         return completedFuture(null);
     }
 
+    @Override
+    public CompletableFuture<List<CompletableFuture<SchemaAndMetadata>>> getAllSchemas(String schemaId) {
+        return completedFuture(null);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/DefaultSchemaRegistryService.java]: `return completedFuture(Collections.emptyList());`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/DefaultSchemaRegistryService.java [hunk 2]
```diff
@@ -47,14 +54,13 @@
     }
 
     @Override
-    public SchemaVersion versionFromBytes(byte[] version) {
-        return null;
+    public CompletableFuture<Boolean> isCompatible(String schemaId, SchemaData schema, SchemaCompatibilityStrategy strategy) {
+        return completedFuture(null);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/DefaultSchemaRegistryService.java]:
```suggestion
        return completedFuture(false);
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java [hunk 3]
```diff
@@ -81,16 +82,32 @@
...
             .thenApply(
                 (existingSchema) ->
-                    existingSchema == null
-                        || existingSchema.schema.isDeleted()
-                        || isCompatible(existingSchema, schema, strategy))
+                {
+                    try {
+                        return existingSchema == null
+                            || existingSchema.schema.isDeleted()
+                            || (isCompatible(schemaId, schema, strategy).get());
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java]: I don't think it is a good idea to use blocking `get` here.

The whole logic can be rewritten as

```
return getSchema(schemaId)
    .thenCompose(
        (existingSchema) -> {
            if (existingSchema == null || existingSchema.schema.isDeleted()) {
                return completedFuture(true);
            } else {
                return isCompatible(schemaId, schema, strategy);
            }
        }
    }
    .thenCompose(...)
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java [hunk 4]
```diff
@@ -162,6 +186,22 @@ private boolean isCompatible(SchemaAndMetadata existingSchema, SchemaData newSch
                         && isCompatible(existingSchema, schema, strategy));
     }
 
+    private CompletableFuture<Boolean> checkCompatibilityWithAll(String schemaId, SchemaData schema,
+                                                                 SchemaCompatibilityStrategy strategy) {
+
+        List<SchemaData> schemas = new ArrayList<>();
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java]:
```bash
getAllSchemas(schemaId).thenCompose(schemaFutures -> {
    return FutureUtils.collect(schemaFutures);
}).thenApply(schemas -> compatibilityChecks.getOrDefault(schema.getType(), SchemaCompatibilityCheck.DEFAULT)
                .isCompatible(schemas, schema, strategy));
```

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by codelipenghui** [comment]
@sijie I'm already address the problem you care about, please take a look.

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by skyrocknroll** [comment]
@codelipenghui Can you please rebase

**Comment by codelipenghui** [comment]
@skyrocknroll @sijie I'm already done the rebase, some changes have been made to #4252. I would like to create a new PR to let client get the reason why schema compatibility validation fails, this requires some improvements to the current `SchemaCompatibilityCheck`

@KannarFr Can you help review these changes? Will this have an impact to #4252 .

**Comment by codelipenghui** [comment]
run cpp Tests

**Pull request merged**